Runtime/JSReference.cs

```diff
@@ -64,11 +89,21 @@ protected virtual void Dispose(bool disposing)
     {
         if (!IsDisposed)
         {
+            // The reference handle should be deleted regardless of the `disposing` argument.
+            napi_status status = napi_delete_reference(_env, _handle);
+
+            // Ignore errors when finalizing.
             if (disposing)
             {
-                napi_delete_reference(_env, _handle).ThrowIfFailed();
```

Comment:
>napi_delete_reference [](http://example.com/codeflow?start=16&length=21)

We are only allowed to call this function from the JS thread. Calling it from the finalizer thread will cause a crash because this API is not thread-safe.

Reply:
Oh, well that's going to be a challenge. I assume the same is true for `napi_get_reference_value` and other things. I guess we'll need to set up a way to dispatch some calls to the JS thread. And then some C# test code that uses marshalled collections from another thread, to detect this kind of issue.

Reply:
I see `napi_threadsafe_function` (added after I left the N-API project), that looks like it can be used for this kind of dispatch. I think I'll make a separate PR to set up that dispatching framework, then incorporate it here afterward.

Reply:
Right. The expectation was that the Dispose method is always called from the JS thread. This why it was always called if disposing == true.

Reply:
Could you remove the call from the finalizer in this PR? The references are deleted on napi_env shutdown. There will be no memory leak. But with this new code there will be a crash.

Reply:
> The expectation was that the Dispose method is always called from the JS thread.

I don't think that's a reasonable assumption. It will be common for some other C# class to have a `JSReference` member, then when that class is disposed it would dispose the reference, possibly from another thread.

As you have said, when we do more async / threading testing we're likely to uncover lots of issues like this. But I _also_ think we will be able to put enough smarts in the NodeAPI library so that developers won't have to worry about it in most cases. I was sort of hoping we could wait on tackling that for just a little while longer though.

Reply:
> Could you remove the call from the finalizer in this PR? The references are deleted on napi_env shutdown.

Yes, I can remove it for now to avoid blocking the rest of this PR.

Reply:
I agree that we may implement a better behavior in future. I am just asking to fix the issue identified in this PR.